libkbfs/tlf_handle_resolve.go

```
@@ -0,0 +1,560 @@
+	nameUIDPair, keybase1.SocialAssertion, error) {
+	nuid, sa, err := ra.resolvableAssertion.resolve(ctx)
+	if err != nil {
+		return nuid, sa, err
+	}
+	if nuid.name.String() != "" {
+		if nuid.name.String() != ra.assertion {
+			sendIfPossible(ra.changed)
+		}
+	} else if sa != (keybase1.SocialAssertion{}) {
+		if sa.String() != ra.assertion {
+			sendIfPossible(ra.changed)
+		}
+	}
+	return nuid, sa, nil
+}
+
+func sendIfPossible(ch chan struct{}) {
+	select {
+	case ch <- struct{}{}:
+	default:
+	}
+}
+
+type resolvableAssertion struct {
+	resolver   resolver
+	assertion  string
+	mustBeUser keybase1.UID
+}
+
+func (ra resolvableAssertion) resolve(ctx context.Context) (
+	nameUIDPair, keybase1.SocialAssertion, error) {
+	if ra.assertion == PublicUIDName {
+		return nameUIDPair{}, keybase1.SocialAssertion{}, fmt.Errorf("Invalid name %s", ra.assertion)
+	}
+	name, uid, err := ra.resolver.Resolve(ctx, ra.assertion)
+	if err == nil && ra.mustBeUser != keybase1.UID("") && ra.mustBeUser != uid {
+		// Force an unresolved assertion sinced the forced user doesn't match
+		err = NoSuchUserError{ra.assertion}
+	}
+	switch err := err.(type) {
+	default:
+		return nameUIDPair{}, keybase1.SocialAssertion{}, err
+	case nil:
+		return nameUIDPair{
+			name: name,
+			uid:  uid,
+		}, keybase1.SocialAssertion{}, nil
+	case NoSuchUserError:
+		socialAssertion, ok := externals.NormalizeSocialAssertion(ra.assertion)
+		if !ok {
+			return nameUIDPair{}, keybase1.SocialAssertion{}, err
+		}
+		return nameUIDPair{}, socialAssertion, nil
+	}
+}
+
+// ParseTlfHandle parses a TlfHandle from an encoded string. See
+// TlfHandle.GetCanonicalName() for the opposite direction.
+//
+// Some errors that may be returned and can be specially handled:
+//
+// TlfNameNotCanonical: Returned when the given name is not canonical
+// -- another name to try (which itself may not be canonical) is in
+// the error. Usually, you want to treat this as a symlink to the name
+// to try.
```

You should comment here that the error is not returned when the ordering is wrong.